DGL sparse SpMM unit test fails on PyTorch 1.12/1.13 + CUDA 11.6

## 🐛 Bug

The current master branch code fails on the following GPU unit test
```
tests/pytorch/mock_sparse2/test_matmul.py::test_spmm[10-1-shape0-rand_csc]
```
with results far off:
```
================================================================================================================================= FAILURES ==================================================================================================================================______________________________________________________________________________________________________________________ test_spmm[10-1-shape0-rand_csc] ______________________________________________________________________________________________________________________

create_func = <function rand_csc at 0x7f8bb0264f28>, shape = (2, 7), nnz = 1, out_dim = 10

    @pytest.mark.parametrize("create_func", [rand_coo, rand_csr, rand_csc])
    @pytest.mark.parametrize("shape", [(2, 7), (5, 2)])
    @pytest.mark.parametrize("nnz", [1, 10])
    @pytest.mark.parametrize("out_dim", [None, 10])
    def test_spmm(create_func, shape, nnz, out_dim):
        dev = F.ctx()
        A = create_func(shape, nnz, dev)
        if out_dim is not None:
            X = torch.randn(shape[1], out_dim, requires_grad=True, device=dev)
        else:
            X = torch.randn(shape[1], requires_grad=True, device=dev)

        sparse_result = A @ X
        grad = torch.randn_like(sparse_result)
        sparse_result.backward(grad)

        adj = sparse_matrix_to_torch_sparse(A)
        XX = clone_detach_and_grad(X)
        torch_sparse_result = torch.sparse.mm(
            adj, XX.view(-1, 1) if out_dim is None else XX
        )
        if out_dim is None:
            torch_sparse_result = torch_sparse_result.view(-1)
        torch_sparse_result.backward(grad)
>       assert torch.allclose(sparse_result, torch_sparse_result, atol=1e-05)
E       AssertionError: assert False
E        +  where False = <built-in method allclose of type object at 0x7f8c32ad1920>(tensor([[ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,\n          0.0000,  0.0000],\n        [...82, -0.3384, -0.9344, -1.9221,  0.2736,  1.6345,\n          2.3735,  2.2325]], device='cuda:0', grad_fn=<SpMMAutoGrad>>), tensor([[ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000,\n          0.0000,  0.0000],\n        [...0648, -3.1924,  2.3995,  9.1691, -6.5909,\n         -1.5264, -2.0847]], device='cuda:0', grad_fn=<SparseAddmmBackward0>), atol=1e-05)
E        +    where <built-in method allclose of type object at 0x7f8c32ad1920> = torch.allclose

tests/pytorch/mock_sparse2/test_matmul.py:48: AssertionError
```

It passes on CPU however.

## To Reproduce

Build master branch code and run GPU unit tests in PyTorch 1.13 + CUDA 11.6.
The docker container `dgllib/dgl-ci-gpu:cu116_v221216` contains the environment I'm testing.
EDIT: Using PyTorch 1.12 will throw the same error.

## Expected behavior

The unit test should pass.

## Environment

 - DGL Version (e.g., 1.0): master
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): PyTorch 1.13.0 + CUDA 11.6.  Same for PyTorch 1.12.
 - OS (e.g., Linux): Linux
 - How you installed DGL (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): `cmake -DUSE_OPENMP=ON -DUSE_CUDA=ON -DBUILD_TORCH=ON -DBUILD_SPARSE=ON`
 - Python version: 3.7.0
 - CUDA/cuDNN version (if applicable): 11.6
 - GPU models and configuration (e.g. V100): T4
 - Any other relevant information:

## Additional context

<!-- Add any other context about the problem here. -->

--- a/tests/pytorch/sparse/test_matmul.py
+++ b/tests/pytorch/sparse/test_matmul.py
@@ -37,18 +37,16 @@ def test_spmm(create_func, shape, nnz, out_dim):
     grad = torch.randn_like(sparse_result)
     sparse_result.backward(grad)
 
-    adj = sparse_matrix_to_torch_sparse(A)
+    adj = sparse_matrix_to_dense(A)
     XX = clone_detach_and_grad(X)
-    torch_sparse_result = torch.sparse.mm(
-        adj, XX.view(-1, 1) if out_dim is None else XX
-    )
+    dense_result = torch.matmul(adj, XX)
     if out_dim is None:
-        torch_sparse_result = torch_sparse_result.view(-1)
-    torch_sparse_result.backward(grad)
-    assert torch.allclose(sparse_result, torch_sparse_result, atol=1e-05)
+        dense_result = dense_result.view(-1)
+    dense_result.backward(grad)
+    assert torch.allclose(sparse_result, dense_result, atol=1e-05)
     assert torch.allclose(X.grad, XX.grad, atol=1e-05)
     assert torch.allclose(
-        adj.grad.coalesce().to_dense(),
+        dense_mask(adj.grad, A),
         sparse_matrix_to_dense(val_like(A, A.val.grad)),
         atol=1e-05,
     )

--- a/tests/pytorch/sparse/utils.py
+++ b/tests/pytorch/sparse/utils.py
@@ -107,8 +107,7 @@ def rand_csc_uncoalesced(shape, nnz, dev):
 
 def sparse_matrix_to_dense(A: SparseMatrix):
     dense = A.dense()
-    dense.requires_grad_()
-    return dense
+    return clone_detach_and_grad(dense)
 
 
 def sparse_matrix_to_torch_sparse(A: SparseMatrix, val=None):

